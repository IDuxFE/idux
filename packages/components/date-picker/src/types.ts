/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { FormSize } from '@idux/components/form'
import type { TimePickerProps } from '@idux/components/time-picker'
import type { DefineComponent, HTMLAttributes, VNode, VNodeTypes } from 'vue'

import { controlPropDef } from '@idux/cdk/forms'
import { ɵPortalTargetDef } from '@idux/cdk/portal'
import { IxPropTypes } from '@idux/cdk/utils'

const datePickerCommonProps = {
  control: controlPropDef,
  open: IxPropTypes.bool,

  allowInput: IxPropTypes.oneOfType([Boolean, IxPropTypes.oneOf(['overlay'])]),
  autofocus: IxPropTypes.bool.def(false),
  borderless: IxPropTypes.bool,
  clearable: IxPropTypes.bool,
  clearIcon: IxPropTypes.string,
  disabled: IxPropTypes.bool.def(false),
  disabledDate: IxPropTypes.func<(date: Date) => boolean>(),
  format: IxPropTypes.string,
  overlayClassName: IxPropTypes.string,
  overlayContainer: IxPropTypes.oneOfType([String, HTMLElement, IxPropTypes.func<() => string | HTMLElement>()]),
  overlayRender: IxPropTypes.func<(children: VNode[]) => VNodeTypes>(),
  readonly: IxPropTypes.bool.def(false),
  size: IxPropTypes.oneOf<FormSize>(['sm', 'md', 'lg']),
  suffix: IxPropTypes.string,
  target: ɵPortalTargetDef,
  type: IxPropTypes.oneOf<DatePickerType>(['date', 'week', 'month', 'quarter', 'year']).def('date'),

  // events
  'onUpdate:open': IxPropTypes.emit<(isOpen: boolean) => void>(),
  onClear: IxPropTypes.emit<(evt: MouseEvent) => void>(),
  onFocus: IxPropTypes.emit<(evt: FocusEvent) => void>(),
  onBlur: IxPropTypes.emit<(evt: FocusEvent) => void>(),
}

export type DatePickerCommonProps = ExtractInnerPropTypes<typeof datePickerCommonProps>
export interface DatePickerCommonBindings {
  blur: () => void
  focus: (options?: FocusOptions) => void
}

export const datePickerProps = {
  ...datePickerCommonProps,

  value: IxPropTypes.oneOfType<number | string | Date>([Number, String, Date]),
  defaultOpenValue: IxPropTypes.oneOfType<number | string | Date>([Number, String, Date]),
  placeholder: IxPropTypes.string,
  timePicker: IxPropTypes.oneOfType([Boolean, IxPropTypes.object<TimePickerProps>()]),

  'onUpdate:value': IxPropTypes.emit<(value: Date) => void>(),
  onChange: IxPropTypes.emit<(value: Date, oldValue: Date) => void>(),
  onInput: IxPropTypes.emit<(evt: Event) => void>(),
}

export type DatePickerProps = ExtractInnerPropTypes<typeof datePickerProps>
export type DatePickerPublicProps = ExtractPublicPropTypes<typeof datePickerProps>
export type DatePickerComponent = DefineComponent<
  Omit<HTMLAttributes, keyof DatePickerPublicProps> & DatePickerPublicProps,
  DatePickerCommonBindings
>
export type DatePickerInstance = InstanceType<DefineComponent<DatePickerProps, DatePickerCommonBindings>>

export const dateRangePickerProps = {
  ...datePickerCommonProps,

  value: IxPropTypes.array<number | string | Date>(),
  defaultOpenValue: IxPropTypes.array<number | string | Date>(),
  placeholder: IxPropTypes.arrayOf(String),
  separator: IxPropTypes.oneOfType([String, IxPropTypes.vNode]),
  timePicker: IxPropTypes.oneOfType([
    Boolean,
    IxPropTypes.object<TimePickerProps>(),
    IxPropTypes.array<TimePickerProps>(),
  ]),

  'onUpdate:value': IxPropTypes.emit<(value: any[]) => void>(),
}

export type DateRangePickerProps = ExtractInnerPropTypes<typeof dateRangePickerProps>
export type DateRangePickerPublicProps = ExtractPublicPropTypes<typeof dateRangePickerProps>
export type DateRangePickerComponent = DefineComponent<
  Omit<HTMLAttributes, keyof DateRangePickerPublicProps> & DateRangePickerPublicProps,
  DatePickerCommonBindings
>
export type DateRangePickerInstance = InstanceType<DefineComponent<DateRangePickerProps, DatePickerCommonBindings>>

export type DatePickerType = 'date' | 'week' | 'month' | 'quarter' | 'year'

// private

export const panelRowProps = {
  rowIndex: IxPropTypes.number.isRequired,
}

export const panelCellProps = {
  rowIndex: IxPropTypes.number.isRequired,
  cellIndex: IxPropTypes.number.isRequired,
  isWeek: IxPropTypes.bool,
}
