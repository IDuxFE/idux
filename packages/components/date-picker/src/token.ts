/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import type { PickerControlContext } from './composables/useControl'
import type { FormatContext } from './composables/useFormat'
import type { InputEnableStatus } from './composables/useInputEnableStatus'
import type { OverlayStateContext } from './composables/useOverlayState'
import type { PickerStateContext } from './composables/usePickerState'
import type { PickerRangeControlContext } from './composables/useRangeControl'
import type { DatePickerProps, DateRangePickerProps } from './types'
import type { CommonConfig, DateConfig, DatePickerConfig } from '@idux/components/config'
import type { Locale } from '@idux/components/locales'
import type { ComputedRef, InjectionKey, Ref, Slots, VNodeTypes } from 'vue'

export interface DatePickerContext extends OverlayStateContext, FormatContext, PickerStateContext<DatePickerProps> {
  props: DatePickerProps
  slots: Slots
  common: CommonConfig
  locale: Locale
  config: DatePickerConfig
  focused: ComputedRef<boolean>
  mergedPrefixCls: ComputedRef<string>
  dateConfig: DateConfig
  inputRef: Ref<HTMLInputElement | undefined>
  inputEnableStatus: ComputedRef<InputEnableStatus>
  controlContext: PickerControlContext
  handleKeyDown: (evt: KeyboardEvent) => void
}

export interface DateRangePickerContext
  extends OverlayStateContext,
    FormatContext,
    PickerStateContext<DateRangePickerProps> {
  props: DateRangePickerProps
  slots: Slots
  common: CommonConfig
  focused: ComputedRef<boolean>
  locale: Locale
  config: DatePickerConfig
  mergedPrefixCls: ComputedRef<string>
  dateConfig: DateConfig
  inputRef: Ref<HTMLInputElement | undefined>
  inputEnableStatus: ComputedRef<InputEnableStatus>
  rangeControlContext: PickerRangeControlContext
  renderSeparator: () => VNodeTypes
  handleKeyDown: (evt: KeyboardEvent) => void
}

export const datePickerToken: InjectionKey<DatePickerContext> = Symbol('datePickerToken')
export const dateRangePickerToken: InjectionKey<DateRangePickerContext> = Symbol('dateRangePickerToken')
