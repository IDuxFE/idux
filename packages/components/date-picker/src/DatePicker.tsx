/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, nextTick, normalizeClass, provide, watch } from 'vue'

import { ɵOverlay } from '@idux/components/_private/overlay'
import { useDateConfig, useGlobalConfig } from '@idux/components/config'
import { useFormElement } from '@idux/components/form'

import { useControl } from './composables/useControl'
import { useFormat } from './composables/useFormat'
import { useInputEnableStatus } from './composables/useInputEnableStatus'
import { useKeyboardEvents } from './composables/useKeyboardEvents'
import { useOverlayProps } from './composables/useOverlayProps'
import { useOverlayState } from './composables/useOverlayState'
import { usePickerState } from './composables/usePickerState'
import Content from './content/Content'
import { datePickerToken } from './token'
import Trigger from './trigger/Trigger'
import { datePickerProps } from './types'

export default defineComponent({
  name: 'IxDatePicker',
  inheritAttrs: false,
  props: datePickerProps,
  setup(props, { attrs, expose, slots }) {
    const common = useGlobalConfig('common')
    const locale = useGlobalConfig('locale')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-date-picker`)
    const config = useGlobalConfig('datePicker')
    const dateConfig = useDateConfig()

    const { elementRef: inputRef, focus, blur } = useFormElement<HTMLInputElement>()

    expose({ focus, blur })

    const inputEnableStatus = useInputEnableStatus(props, config)
    const formatContext = useFormat(props, config)
    const pickerStateContext = usePickerState(props, dateConfig, formatContext.formatRef)

    const { accessor, handleChange } = pickerStateContext

    const controlContext = useControl(dateConfig, formatContext, inputEnableStatus, accessor.valueRef, handleChange)
    const { overlayOpened, setOverlayOpened } = useOverlayState(props, controlContext)
    const handleKeyDown = useKeyboardEvents(setOverlayOpened)

    const context = {
      props,
      slots,
      locale,
      config,
      mergedPrefixCls,
      dateConfig,
      inputRef,
      inputEnableStatus,
      overlayOpened,
      setOverlayOpened,
      handleKeyDown,
      controlContext,
      ...formatContext,
      ...pickerStateContext,
    }

    provide(datePickerToken, context)

    watch(overlayOpened, opened => {
      nextTick(() => {
        if (opened) {
          focus()
          inputRef.value?.dispatchEvent(new FocusEvent('focus'))
        } else {
          blur()
          inputRef.value?.dispatchEvent(new FocusEvent('blur'))
        }
      })
    })

    const renderTrigger = () => <Trigger {...attrs}></Trigger>
    const renderContent = () => <Content></Content>
    const overlayProps = useOverlayProps(context)
    const overlayClass = computed(() => normalizeClass([`${mergedPrefixCls.value}-overlay`, props.overlayClassName]))

    return () => (
      <ɵOverlay
        {...overlayProps.value}
        class={overlayClass.value}
        v-slots={{ default: renderTrigger, content: renderContent }}
        triggerId={attrs.id}
      />
    )
  },
})
