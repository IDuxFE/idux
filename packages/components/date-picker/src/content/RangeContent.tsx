/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, inject, onMounted, onUpdated, ref } from 'vue'

import { convertArray } from '@idux/cdk/utils'
import { ɵDatePanel } from '@idux/components/_private/date-panel'
import { ɵFooter } from '@idux/components/_private/footer'
import { ɵInput, type ɵInputInstance } from '@idux/components/_private/input'
import { ɵTimePanel } from '@idux/components/_private/time-panel'

import { useRangeActiveDate } from '../composables/useActiveDate'
import { dateRangePickerToken } from '../token'

export default defineComponent({
  setup() {
    const context = inject(dateRangePickerToken)!
    const {
      props,
      config,
      dateConfig,
      locale,
      slots,
      overlayOpened,
      formatRef,
      dateFormatRef,
      timeFormatRef,
      rangeControlContext: {
        buffer,
        panelValue,
        visiblePanel,
        isSelecting,
        fromControl,
        toControl,
        handleDatePanelCellClick,
        handleDatePanelCellMouseenter,
      },
      mergedPrefixCls,
      inputEnableStatus,
      inputRef,
      handleChange,
      handleKeyDown,
      renderSeparator,
      setOverlayOpened,
    } = context

    const timePanelProps = computed(() => convertArray(context.props.timePanelOptions))
    const { fromActiveDate, toActiveDate, setFromActiveDate, setToActiveDate } = useRangeActiveDate(
      dateConfig,
      props,
      panelValue,
      isSelecting,
      formatRef,
    )

    const inputInstance = ref<ɵInputInstance>()
    onMounted(() => {
      inputRef.value = inputInstance.value?.getInputElement()
    })
    onUpdated(() => {
      inputRef.value = inputInstance.value?.getInputElement()
    })

    const handleConfirm = () => {
      handleChange(buffer.value)
      setOverlayOpened(false)
    }

    const handleMouseDown = (e: MouseEvent) => {
      if (!(e.target instanceof HTMLInputElement)) {
        e.preventDefault()
      }
    }

    const renderBoard = (prefixCls: string, isFrom: boolean) => {
      const { enableOverlayDateInput, enableOverlayTimeInput } = inputEnableStatus.value

      const {
        dateInputValue,
        timeInputVaue,
        handleDateInput,
        handleTimeInput,
        handleDateInputClear,
        handleTimeInputClear,
        handleDateInputFocus,
        handleTimeInputFocus,
        handleTimePanelChange,
      } = isFrom ? fromControl : toControl

      const inputs = enableOverlayDateInput && (
        <div class={`${prefixCls}-inputs`}>
          <ɵInput
            ref={inputInstance}
            class={`${prefixCls}-date-input`}
            v-slots={slots}
            value={dateInputValue.value}
            size="sm"
            clearable={props.clearable ?? config.clearable}
            clearIcon={props.clearIcon ?? config.clearIcon}
            placeholder={dateFormatRef.value}
            onInput={handleDateInput}
            onFocus={handleDateInputFocus}
            onKeydown={handleKeyDown}
            onClear={handleDateInputClear}
          />
          {enableOverlayTimeInput && (
            <ɵInput
              class={`${prefixCls}-time-input`}
              v-slots={slots}
              value={timeInputVaue.value}
              size="sm"
              clearable={props.clearable ?? config.clearable}
              clearIcon={props.clearIcon ?? config.clearIcon}
              placeholder={timeFormatRef.value}
              onInput={handleTimeInput}
              onFocus={handleTimeInputFocus}
              onKeydown={handleKeyDown}
              onClear={handleTimeInputClear}
            />
          )}
        </div>
      )

      const datePanelProps = {
        cellTooltip: props.cellTooltip,
        disabledDate: props.disabledDate,
        type: props.type === 'datetime' ? 'date' : props.type,
        value: panelValue.value,
        visible: overlayOpened.value,
        activeDate: isFrom ? fromActiveDate.value : toActiveDate.value,
        onCellClick: handleDatePanelCellClick,
        onCellMouseenter: handleDatePanelCellMouseenter,
        'onUpdate:activeDate': isFrom ? setFromActiveDate : setToActiveDate,
      }

      return (
        <div class={prefixCls}>
          {inputs}
          <div class={`${prefixCls}-panel`}>
            <ɵDatePanel v-show={visiblePanel.value !== 'timePanel'} v-slots={slots} {...datePanelProps} />
            {inputEnableStatus.value.enableOverlayTimeInput && (
              <ɵTimePanel
                v-show={visiblePanel.value === 'timePanel'}
                {...(timePanelProps.value[isFrom ? 0 : 1] ?? {})}
                value={panelValue.value?.[isFrom ? 0 : 1]}
                visible={visiblePanel.value === 'timePanel'}
                onChange={handleTimePanelChange}
              />
            )}
          </div>
        </div>
      )
    }

    return () => {
      const prefixCls = `${mergedPrefixCls.value}-overlay`
      const boardPrefixCls = `${mergedPrefixCls.value}-board`

      const children = [
        <div class={`${prefixCls}-content`} tabindex={-1} onMousedown={handleMouseDown}>
          {renderBoard(boardPrefixCls, true)}
          <div class={`${prefixCls}-gap`}>{inputEnableStatus.value.enableOverlayDateInput && renderSeparator()}</div>
          {renderBoard(boardPrefixCls, false)}
        </div>,
        <ɵFooter
          v-slots={slots}
          class={`${prefixCls}-footer`}
          footer={props.footer}
          okText={locale.dateRangePicker.okText}
          okButton={{ size: 'xs', mode: 'primary' }}
          cancelVisible={false}
          ok={handleConfirm}
        />,
      ]

      return props.overlayRender ? (
        props.overlayRender(children)
      ) : (
        <div class={prefixCls} onMousedown={handleMouseDown}>
          {children}
        </div>
      )
    }
  },
})
