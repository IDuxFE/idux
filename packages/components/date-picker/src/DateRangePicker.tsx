/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, normalizeClass, provide, watch } from 'vue'

import { ɵOverlay } from '@idux/components/_private/overlay'
import { useDateConfig, useGlobalConfig } from '@idux/components/config'
import { useFormElement } from '@idux/components/utils'

import { useFormat } from './composables/useFormat'
import { useInputEnableStatus } from './composables/useInputEnableStatus'
import { useRangeKeyboardEvents } from './composables/useKeyboardEvents'
import { useOverlayState } from './composables/useOverlayState'
import { usePickerState } from './composables/usePickerState'
import { useRangeControl } from './composables/useRangeControl'
import RangeContent from './content/RangeContent'
import { dateRangePickerToken } from './token'
import RangeTrigger from './trigger/RangeTrigger'
import { dateRangePickerProps } from './types'

const defaultOffset: [number, number] = [0, 8]

export default defineComponent({
  name: 'IxDateRangePicker',
  inheritAttrs: false,
  props: dateRangePickerProps,
  setup(props, { attrs, expose, slots }) {
    const common = useGlobalConfig('common')
    const locale = useGlobalConfig('locale')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-date-range-picker`)
    const config = useGlobalConfig('datePicker')
    const dateConfig = useDateConfig()

    const { elementRef: inputRef, focus, blur } = useFormElement<HTMLInputElement>()

    expose({ focus, blur })

    const showFooter = computed(() => !!props.footer || !!slots.footer)

    const inputEnableStatus = useInputEnableStatus(props, config)
    const formatContext = useFormat(props, config)
    const pickerStateContext = usePickerState(props, dateConfig, formatContext.formatRef)

    const { accessor, handleChange } = pickerStateContext

    const rangeControlContext = useRangeControl(dateConfig, formatContext, inputEnableStatus, accessor.valueRef)
    const { overlayOpened, setOverlayOpened } = useOverlayState(props, rangeControlContext)
    const handleKeyDown = useRangeKeyboardEvents(rangeControlContext, showFooter, setOverlayOpened, handleChange)

    const renderSeparator = () => slots.separator?.() ?? props.separator ?? locale.dateRangePicker.separator

    provide(dateRangePickerToken, {
      props,
      slots,
      locale,
      config,
      mergedPrefixCls,
      dateConfig,
      inputRef,
      inputEnableStatus,
      overlayOpened,
      setOverlayOpened,
      rangeControlContext,
      renderSeparator,
      handleKeyDown,
      ...formatContext,
      ...pickerStateContext,
    })

    watch(overlayOpened, opened => {
      if (!opened) {
        rangeControlContext.init(true)
      }

      opened ? focus() : blur()
    })

    watch(rangeControlContext.buffer, value => {
      if (!showFooter.value) {
        handleChange(value)
      }
    })

    const target = computed(() => props.target ?? config.target ?? `${mergedPrefixCls.value}-overlay-container`)
    const renderTrigger = () => <RangeTrigger {...attrs}></RangeTrigger>
    const renderContent = () => <RangeContent></RangeContent>
    const overlayProps = { 'onUpdate:visible': setOverlayOpened }

    return () => {
      return (
        <ɵOverlay
          {...overlayProps}
          visible={overlayOpened.value}
          v-slots={{ default: renderTrigger, content: renderContent }}
          class={normalizeClass(props.overlayClassName)}
          clickOutside
          disabled={accessor.disabled.value || props.readonly}
          offset={defaultOffset}
          placement="bottomStart"
          target={target.value}
          trigger="manual"
        />
      )
    }
  },
})
