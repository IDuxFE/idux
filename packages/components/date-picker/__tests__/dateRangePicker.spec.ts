import { MountingOptions, VueWrapper, mount } from '@vue/test-utils'

import { renderWork } from '@tests'
import { parse } from 'date-fns'

import DatePanel from '@idux/components/_private/date-panel/src/DatePanel'

import DatePanelCell from '../../_private/date-panel/src/panel-body/PanelCell'
import { ɵTimePanelInstance } from '../../_private/time-panel'
import TimePanel from '../../_private/time-panel/src/TimePanel'
import TimePanelCell from '../../_private/time-panel/src/TimePanelCell'
import TimePanelColumn from '../../_private/time-panel/src/TimePanelColumn'
import DateRangePicker from '../src/DateRangePicker'
import RangeContent from '../src/content/RangeContent'
import { DateRangePickerInstance, DateRangePickerProps } from '../src/types'

describe('DateRangePicker', () => {
  const DateRangePickerMount = (options?: MountingOptions<Partial<DateRangePickerProps>>) => {
    const { props, ...rest } = options || {}
    return mount(DateRangePicker, {
      props: { open: true, ...props },
      ...(rest as MountingOptions<DateRangePickerProps>),
      attachTo: 'body',
    }) as VueWrapper<DateRangePickerInstance>
  }

  renderWork<DateRangePickerProps>(DateRangePicker, {
    props: {},
  })

  const findTimePanelColumns = (wrapper: VueWrapper<ɵTimePanelInstance>) => wrapper.findAllComponents(TimePanelColumn)
  const findTimePanelColumn = (wrapper: VueWrapper<ɵTimePanelInstance>, idx: number) =>
    findTimePanelColumns(wrapper)?.[idx]

  const findTimePanelCellWithValue = (wrapper: ReturnType<typeof findTimePanelColumn>, value: string | number) =>
    wrapper.findAllComponents(TimePanelCell).find(comp => comp.props().value === value)

  const findTimePanelCell = (wrapper: VueWrapper<ɵTimePanelInstance>, idx: number, value: string | number) =>
    findTimePanelCellWithValue(findTimePanelColumn(wrapper, idx), value)

  const triggerConfirm = (wrapper: VueWrapper<DateRangePickerInstance>) =>
    wrapper
      .findComponent(RangeContent)
      .findAll('.ix-date-range-picker-overlay-footer .ix-button')
      .find(btn => btn.text() === '确定')
      ?.trigger('click')

  const findCell = (wrapper: VueWrapper<DateRangePickerInstance>, type: 'from' | 'to', cellLabel: string) =>
    wrapper
      .findAllComponents(DatePanel)
      [type === 'from' ? 0 : 1].findAllComponents(DatePanelCell)
      .find(cell => cell.find('.ix-date-panel-cell-trigger').text() === cellLabel)

  test('v-model:value work', async () => {
    const onUpdateValue = vi.fn()
    const onChange = vi.fn()
    const onUpdateVisible = vi.fn()
    const wrapper = DateRangePickerMount({
      props: {
        value: [new Date('2021-10-01 00:00:00'), new Date('2021-11-11 00:00:00')],
        'onUpdate:value': onUpdateValue,
        onChange,
        'onUpdate:open': onUpdateVisible,
      },
    })

    expect(findCell(wrapper, 'from', '1')?.classes()).toContain('ix-date-panel-cell-selected')
    expect(findCell(wrapper, 'to', '11')?.classes()).toContain('ix-date-panel-cell-selected')

    await findCell(wrapper, 'from', '11')?.trigger('click')
    await findCell(wrapper, 'to', '21')?.trigger('click')
    await triggerConfirm(wrapper)

    expect(onUpdateValue).toBeCalledWith([new Date('2021-10-11 00:00:00'), new Date('2021-11-21 00:00:00')])
    expect(onChange).toBeCalledWith(
      [new Date('2021-10-11 00:00:00'), new Date('2021-11-21 00:00:00')],
      [new Date('2021-10-01 00:00:00'), new Date('2021-11-11 00:00:00')],
    )
    expect(onUpdateVisible).toBeCalledWith(false)

    await wrapper.setProps({ value: [new Date('2021-9-03 00:00:00'), new Date('2021-12-22 00:00:00')] })
    expect(findCell(wrapper, 'from', '3')?.classes()).toContain('ix-date-panel-cell-selected')
    expect(findCell(wrapper, 'to', '22')?.classes()).toContain('ix-date-panel-cell-selected')
    expect(findCell(wrapper, 'from', '11')?.classes()).toContain('ix-date-panel-cell-in-range')
    expect(findCell(wrapper, 'to', '11')?.classes()).toContain('ix-date-panel-cell-in-range')
  })

  test('v-model:open work', async () => {
    const onUpdateOpen = vi.fn()
    const wrapper = DateRangePickerMount({ props: { open: false, 'onUpdate:open': onUpdateOpen } })

    expect(wrapper.findComponent(RangeContent).exists()).toBeFalsy()

    await wrapper.find('.ix-date-range-picker').trigger('click')
    expect(onUpdateOpen).toBeCalledWith(true)

    await wrapper.setProps({ open: true })
    expect(wrapper.findComponent(RangeContent).isVisible()).toBeTruthy()
  })

  test('autoFocus work', async () => {
    const onUpdateOpen = vi.fn()
    DateRangePickerMount({ props: { open: false, autofocus: true, 'onUpdate:open': onUpdateOpen } })

    expect(onUpdateOpen).toBeCalledWith(true)
  })

  test('disabled work', async () => {
    const onUpdateOpen = vi.fn()
    const onInput = vi.fn()
    const wrapper = DateRangePickerMount({
      props: { open: false, disabled: true, 'onUpdate:open': onUpdateOpen, onInput },
    })

    await wrapper.find('.ix-date-range-picker').trigger('click')
    expect(onUpdateOpen).not.toBeCalled()

    await wrapper.find('.ix-date-range-picker').find('input').setValue('2021-03-01')
    expect(onInput).not.toBeCalled()
  })

  test('clear work', async () => {
    const onUpdateValue = vi.fn()
    const onChange = vi.fn()
    const onClear = vi.fn()
    const wrapper = DateRangePickerMount({
      props: {
        clearable: true,
        value: [new Date('2021-10-01 00:00:00'), new Date('2021-11-11 00:00:00')],
        'onUpdate:value': onUpdateValue,
        onChange,
        onClear,
      },
    })

    await wrapper.find('.ix-trigger-clear-icon').trigger('click')
    expect(onClear).toBeCalled()
    expect(onChange).toBeCalledWith(undefined, [new Date('2021-10-01 00:00:00'), new Date('2021-11-11 00:00:00')])
    expect(onUpdateValue).toBeCalledWith(undefined)
  })

  test('format work', async () => {
    const wrapper = DateRangePickerMount({
      props: { value: [new Date('2021-10-01 00:00:00'), new Date('2021-11-11 00:00:00')], format: 'dd/MM/yyyy' },
    })

    expect(wrapper.find('.ix-date-range-picker').findAll('input')[0].element.value).toBe('01/10/2021')
    expect(wrapper.find('.ix-date-range-picker').findAll('input')[1].element.value).toBe('11/11/2021')
  })

  test('input work', async () => {
    const onInput = vi.fn()
    const onChange = vi.fn()
    const onUpdateValue = vi.fn()
    const wrapper = DateRangePickerMount({
      props: {
        value: [new Date('2021-10-01 00:00:00'), new Date('2021-11-11 00:00:00')],
        format: 'yyyy-MM-dd',
        onInput,
        onChange,
        'onUpdate:value': onUpdateValue,
      },
    })

    await wrapper.find('.ix-date-range-picker').findAll('input')[0].setValue('2021-10-11')
    expect(onInput).toBeCalled()

    onInput.mockClear()

    await wrapper.find('.ix-date-range-picker').findAll('input')[1].setValue('2021-12-22')
    expect(onInput).toBeCalled()

    await triggerConfirm(wrapper)

    const newFromDate = parse('2021-10-11', 'yyyy-MM-dd', new Date())
    const newToDate = parse('2021-12-22', 'yyyy-MM-dd', new Date())
    expect(onChange).toBeCalledWith(
      [newFromDate, newToDate],
      [new Date('2021-10-01 00:00:00'), new Date('2021-11-11 00:00:00')],
    )
    expect(onUpdateValue).toBeCalledWith([newFromDate, newToDate])
  })

  test('defaultOpenValue work', async () => {
    const wrapper = DateRangePickerMount({
      props: {
        value: undefined,
        defaultOpenValue: [new Date('2021-10-11 00:00:00'), new Date('2021-11-11 00:00:00')],
      },
    })

    const [fromDatePanel, toDatePanel] = wrapper.findAllComponents(DatePanel)
    const fromHeaderContentBtns = fromDatePanel.find('.ix-date-panel-header-content').findAll('button')
    const toHeaderContentBtns = toDatePanel.find('.ix-date-panel-header-content').findAll('button')
    expect(fromHeaderContentBtns.some(btn => btn.text().indexOf('2021') > -1)).toBeTruthy()
    expect(fromHeaderContentBtns.some(btn => btn.text().indexOf('10') > -1)).toBeTruthy()
    expect(toHeaderContentBtns.some(btn => btn.text().indexOf('2021') > -1)).toBeTruthy()
    expect(toHeaderContentBtns.some(btn => btn.text().indexOf('11') > -1)).toBeTruthy()
  })

  test('datetime time select work', async () => {
    const onChange = vi.fn()
    const onUpdateValue = vi.fn()
    const wrapper = DateRangePickerMount({
      props: {
        type: 'datetime',
        value: [new Date('2021-10-11 00:00:00'), new Date('2021-11-11 00:00:00')],
        onChange,
        'onUpdate:value': onUpdateValue,
      },
    })

    await wrapper.findComponent(RangeContent).findAll('.ix-date-range-picker-board-time-input')[0].trigger('focus')

    const fromTimePanel = wrapper.findAllComponents(TimePanel)[0] as unknown as VueWrapper<ɵTimePanelInstance>
    const toTimePanel = wrapper.findAllComponents(TimePanel)[1] as unknown as VueWrapper<ɵTimePanelInstance>

    await findTimePanelCell(fromTimePanel, 0, 13)?.trigger('click')
    await findTimePanelCell(fromTimePanel, 1, 3)?.trigger('click')
    await findTimePanelCell(fromTimePanel, 2, 4)?.trigger('click')

    await findTimePanelCell(toTimePanel, 0, 14)?.trigger('click')
    await findTimePanelCell(toTimePanel, 1, 5)?.trigger('click')
    await findTimePanelCell(toTimePanel, 2, 6)?.trigger('click')

    await triggerConfirm(wrapper)

    expect(onUpdateValue).toBeCalledWith([new Date('2021-10-11 13:03:04'), new Date('2021-11-11 14:05:06')])
    expect(onChange).toBeCalledWith(
      [new Date('2021-10-11 13:03:04'), new Date('2021-11-11 14:05:06')],
      [new Date('2021-10-11 00:00:00'), new Date('2021-11-11 00:00:00')],
    )
  })

  test('datetime input work', async () => {
    const onChange = vi.fn()
    const onUpdateValue = vi.fn()
    const wrapper = DateRangePickerMount({
      props: {
        type: 'datetime',
        value: [new Date('2021-10-11 00:00:00'), new Date('2021-11-11 00:00:00')],
        format: 'yyyy-MM-dd HH:mm:ss',
        onChange,
        'onUpdate:value': onUpdateValue,
      },
    })

    const dateInputs = wrapper
      .findComponent(RangeContent)
      .findAll('.ix-date-range-picker-board-date-input')
      .map(el => el.find('input'))
    const timeInputs = wrapper
      .findComponent(RangeContent)
      .findAll('.ix-date-range-picker-board-time-input')
      .map(el => el.find('input'))
    await dateInputs[0].setValue('2021-11-22')
    await dateInputs[1].setValue('2021-12-25')
    await timeInputs[0].setValue('13:03:04')
    await timeInputs[1].setValue('14:05:06')

    await triggerConfirm(wrapper)

    expect(onUpdateValue).toBeCalledWith([new Date('2021-11-22 13:03:04'), new Date('2021-12-25 14:05:06')])
    expect(onChange).toBeCalledWith(
      [new Date('2021-11-22 13:03:04'), new Date('2021-12-25 14:05:06')],
      [new Date('2021-10-11 00:00:00'), new Date('2021-11-11 00:00:00')],
    )

    onUpdateValue.mockClear()
    onChange.mockClear()

    const triggerInputs = wrapper.find('.ix-date-range-picker').findAll('input')
    await triggerInputs[0].setValue('2021-11-22 13:03:04')
    await triggerInputs[1].setValue('2021-12-25 14:05:06')

    await triggerConfirm(wrapper)

    expect(onUpdateValue).toBeCalledWith([new Date('2021-11-22 13:03:04'), new Date('2021-12-25 14:05:06')])
    expect(onChange).toBeCalledWith(
      [new Date('2021-11-22 13:03:04'), new Date('2021-12-25 14:05:06')],
      [new Date('2021-10-11 00:00:00'), new Date('2021-11-11 00:00:00')],
    )
  })
})
