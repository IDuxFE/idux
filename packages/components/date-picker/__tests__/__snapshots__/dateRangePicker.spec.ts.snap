// Vitest Snapshot v1

exports[`DateRangePicker > render work 1`] = `
"<div class=\\"ix-date-range-picker ix-trigger ix-trigger-md ix-trigger-readonly\\">
  <div class=\\"ix-date-range-picker-input\\"><input class=\\"ix-date-range-picker-input-inner\\" autocomplete=\\"off\\" placeholder=\\"开始日期\\" readonly=\\"\\" size=\\"12\\"><span class=\\"ix-date-range-picker-input-separator\\">至</span><input class=\\"ix-date-range-picker-input-inner\\" autocomplete=\\"off\\" placeholder=\\"结束日期\\" readonly=\\"\\" size=\\"12\\"></div>
  <div class=\\"ix-trigger-suffix\\"><i class=\\"ix-icon ix-icon-calendar\\" role=\\"img\\" aria-label=\\"calendar\\"></i></div>
  <!---->
</div>"
`;
